File: gearshift2/src/main/java/org/sugr/gearshift/viewmodel/RetainedViewModel.java
package org.sugr.gearshift.viewmodel;

import android.content.SharedPreferences;
import android.support.annotation.Nullable;

import com.f2prateek.rx.preferences.RxSharedPreferences;

import rx.Observable;
import rx.subjects.PublishSubject;

public class RetainedViewModel<T> {
    protected final RxSharedPreferences prefs;
    @Nullable protected T consumer;

    private PublishSubject<Void> unbindSubject = PublishSubject.create();
    private PublishSubject<Void> destroySubject = PublishSubject.create();

    public RetainedViewModel(SharedPreferences prefs) {
        this.prefs = RxSharedPreferences.create(prefs);
    }

    public void bind(T consumer) {
        this.consumer = consumer;
    }

    public void unbind() {
        unbindSubject.onNext(null);

        consumer = null;
    }

    public void onDestroy() {
        destroySubject.onNext(null);
    }

    // Useful for stopping observable emission when an unbind happens
    public <O> Observable.Transformer<O, O> takeUntilUnbind() {
        return o -> o.takeUntil(unbindSubject);
    }

    // Similar to takeUntilUnbind, but for onDestroy
    public <O> Observable.Transformer<O, O> takeUntilDestroy() {
        return o -> o.takeUntil(destroySubject);
    }
}

File: gearshift2/src/main/kotlin/org/sugr/gearshift/viewmodel/RetainedViewModel.kt
package org.sugr.gearshift.viewmodel

import android.content.SharedPreferences

import com.f2prateek.rx.preferences.RxSharedPreferences
import org.sugr.gearshift.logD
import org.sugr.gearshift.viewmodel.rxutil.sharedPreferences

import rx.Observable
import rx.lang.kotlin.PublishSubject

open class RetainedViewModel<T>(val tag: String, prefs: SharedPreferences) {
    protected val prefs: RxSharedPreferences
    protected var consumer: T? = null

    private val unbindSubject = PublishSubject<Void>()
    private val destroySubject = PublishSubject<Void>()

    init {
        logD("Creating $this view model")
        this.prefs = sharedPreferences(prefs)
    }

    open fun bind(consumer: T) {
        logD("Binding $this view model")
        this.consumer = consumer
    }

    fun unbind() {
        logD("Unbinding $this view model")
        unbindSubject.onNext(null)

        consumer = null
    }

    fun onDestroy() {
        logD("Destroying $this view model")
        destroySubject.onNext(null)
    }

    // Useful for stopping observable emission when an unbind happens
    fun <O> takeUntilUnbind(): Observable.Transformer<O, O> {
        return Observable.Transformer { o -> o.takeUntil(unbindSubject) }
    }

    // Similar to takeUntilUnbind, but for onDestroy
    fun <O> takeUntilDestroy(): Observable.Transformer<O, O> {
        return Observable.Transformer { o -> o.takeUntil(destroySubject) }
    }
}

interface LeaveBlocker {
    fun canLeave(): Boolean
}